analysis/dataset_definition_dm_reg.py

```diff
@@ -0,0 +1,86 @@
+from datetime import date, timedelta
+
+from databuilder.ehrql import Dataset, case, when
+from databuilder.tables.beta.tpp import (
+    clinical_events,
+    patients,
+    practice_registrations,
+)
+
+import codelists
+from variable_lib_helper import (
+    first_matching_event,
+    last_matching_event,
+    age_as_of,
+    died_as_of,
+    practice_registration_as_of,
+)
+
+# Payment Period End Date (PPED)
+index_date = date(2022, 3, 31)
+
+# Create dataset
+dataset = Dataset()
+
+# Define clinical events prior to index_date
+prior_events = clinical_events.take(clinical_events.date.is_on_or_before(index_date))
+
+# Define variable indicating whether a patient has died
+# Note: Although this is not in the business rules I'm including this as an
+# additional check
+has_not_died = ~died_as_of(index_date)
+
+
+# GMS reg status: Select patients who meet either of the criteria below:
+# Registered for GMS prior to or on the achievement date and did not
+# subsequently deregister from GMS (currently registered for GMS).
+# Registered for GMS prior to or on the achievement date and subsequently
+# deregistered from GMS after the achievement date (previously registered for
+# GMS). (i.e. patients who were registered for GMS on the achievement date).
+# TODO Note that this currently does not follow the qualifying criteria in the
+# business rules.
+practice_reg = practice_registration_as_of(index_date)
+gms_registration_status = practice_reg.exists_for_patient()
+
+# Define clinical data extraction criteria
+
+# Field number 4
+# PAT_AGE: The age of the patient in full years at the achievement date.
+dataset.pat_age = age_as_of(index_date)
+
+# Field number 6
+# DMLAT_DAT: Date of the most recent diabetes diagnosis up to and including the
+# achievement date.
+dataset.dm_lat_dat = last_matching_event(prior_events, codelists.dm_cod).date
+
+# Field number 7
+# DMRES_DAT: Date of the most recent diabetes diagnosis resolved code recorded
+# after the most recent diabetes diagnosis and up to and including the
+# achievement date.
+dataset.dm_res_dat = last_matching_event(prior_events, codelists.dm_res_cod).date
+
+########################################################
+# Define Business Rules for Diabetes register (RM_REG) #
+########################################################
+
+# DM_REG: Rule number 1:
+# Pass to the next rule all patients from the specified population who meet
+# both of the criteria below:  Have a diabetes diagnosis in the patient record
+# up to and including the achievement date. Latest diabetes diagnosis is not
+# followed by a diabetes resolved code.
+dm_reg_r1 = (dataset.dm_res_dat < dataset.dm_lat_dat) | (
```

Comment:
yes, that's what this is and a much clearer description. 

the main reason I decided to stick to the source names here is because I find it clearer when setting the population. for the diabetes register it doesnt seem that important because (1) it is easy to describe the rules with short names and (2) there are only 2 rules. other indicators have many more rules. for example `dataset.set_population` for indicator DM020 (10 rules) would look something like this:

```
dataset.set_population(
    dm_register
    & dm020_r1
    & dm020_r2
    & dm020_r3
    & dm020_r4
    & dm020_r5
    & dm020_r6
    & dm020_r7
    & dm020_r8
    & dm020_r9
    & dm020_r10
)
```

if we had descriptive names for each rule we'd probably struggle to find short names for very complex rules. e.g., rule number 8:

<img width="946" alt="Screenshot 2022-12-09 at 16 31 29" src="https://user-images.githubusercontent.com/25428744/206748292-ec72eb7e-43f1-4474-b11e-9d590664a678.png">

at the same time I agree that `dm_reg_r1` doesnt describe at all what is going and is meaningless without looking it up in the QOF business rules (word document only available online in a .zip) or in the code comments.